### PR TITLE
CI: run integration tests against all supported Kubernetes versions

### LINKED ISSUE #3396: testing: run integration tests against all supported Kubernetes versions
Per https://projectcontour.io/resources/compatibility-matrix/, each version of Contour is considered compatible with three minor Kubernetes versions. However, our integration tests tend to run only against the latest version (based on whatever image the version of kind that we're using pulls by default).

We should add runs of the tests against the other two versions as well. Personally, I'd be OK with this starting as a nightly job, as I don't know that we *need* to run against all versions for every PR. Open to other inputs on this though.

### PR DESCRIPTION
**Pull request by skriss**
Runs integration tests against the three supported minor versions
of Kubernetes.

Closes #3396.

Signed-off-by: Steve Kriss <krisss@vmware.com>

### CODE REVIEW

**Comment by skriss** [comment]
This is the quick and dirty version of this. On the plus side, integration tests are run against each minor version in parallel, so it won't slow down CI for PRs. On the downside, the versions to test against are hardcoded and will have to be periodically updated. @projectcontour/maintainers WDYT? Good enough for now?

**Comment by stevesloka** [comment]
Looks fine to me! I don't know of a good way around the hard coding unless we somehow could tie it to the support matrix so an update there updated this, but for now it's a one every three month change.

**Comment by skriss** [comment]
> Looks fine to me! I don't know of a good way around the hard coding unless we somehow could tie it to the support matrix so an update there updated this, but for now it's a one every three month change.

Yeah, I agree, it seemed not worth the effort to try to automate. Can always revisit if we need to.

**Review by youngnick — APPROVED** [review]
LGTM, I think automating this more at the moment is probably premature. Nice one.

**Comment by skriss** [comment]
will require a branch protection rules update when merged to account for the three new required job names and plain old `integration` being dropped.

**Review by sunjayBhatia — APPROVED** [review]

File: .github/workflows/prbuild.yaml [hunk 1]
```diff
@@ -6,8 +6,36 @@ on: [push, pull_request]
 env:
   GOPROXY: https://proxy.golang.org/
 jobs:
-  integration:
+  integration-latest:
     runs-on: ubuntu-latest
+    env:
+      NODEIMAGE: docker.io/kindest/node:v1.20.2
+    steps:
```
Comment (sunjayBhatia) [inline .github/workflows/prbuild.yaml:13]: Oh just occurred to me could we use YAML anchors/references to DRY this out?

Comment (skriss) [reply .github/workflows/prbuild.yaml:13]: did some googling, AFAICT GH Actions does not support YAML anchors: https://github.community/t/support-for-yaml-anchors/16128

Comment (sunjayBhatia) [reply .github/workflows/prbuild.yaml:13]: ah thats no good :(

**Review by stevesloka — APPROVED** [review]

File: .github/workflows/prbuild.yaml (hunk 1 above)
Comment (sunjayBhatia) [reply .github/workflows/prbuild.yaml:13]: maybe we could use this? https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstrategymatrix

Comment (sunjayBhatia) [reply .github/workflows/prbuild.yaml:13]: could make it so we only have 1 "job" still but different steps within that

Comment (skriss) [reply .github/workflows/prbuild.yaml]: Switched to a matrix and it runs well. The only annoying thing is that now in branch protection rules, the required job names include the specific node image value, so we'll have to change those every time we change a Kube version we run against. Looking to see if there's a way around that..

Comment (sunjayBhatia) [reply .github/workflows/prbuild.yaml]: oh that is very annoying 👎🏽

File: .github/workflows/prbuild.yaml [hunk 2]
```diff
@@ -8,13 +8,31 @@ env:
 jobs:
   integration:
     runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        # use stable kubernetes_version values since they're included
```
Comment (sunjayBhatia) [inline .github/workflows/prbuild.yaml:13]: pretty slick, nice!

Comment (skriss) [reply .github/workflows/prbuild.yaml:13]: :+1: I think this is the best of both worlds, thanks for pointing me in this direction. Now if only the jobs will run...

**Pull request merged**